Add Login and Logout feature

aditya102 opened the pull request:
- Added Login, Logout and Registration pages
 - Updated layout with crispy_form
- Added testcase for Login and Logout

waranlogesh commented on tests/test_user/test_login.py:
```diff
@@ -0,0 +1,57 @@
+        response = self.client.get(reverse('login'))
+        self.assertEqual(response.status_code, 200)
+        self.assertTemplateUsed(response, 'users/login.html')
+        self.assertContains(response, '<div id="div_id_username" class="form-group"> <label for="id_username" class="col-form-label  requiredField">')
```
flake8.

waranlogesh commented on tests/test_user/test_login.py:
```diff
@@ -0,0 +1,57 @@
+
+class AccountsTestCase(TestCase):
+    def setUp(self):
+        user = {
```
No need for this.

aditya102 replied:
removed

waranlogesh commented on tests/test_user/test_login.py:
```diff
@@ -0,0 +1,57 @@
+        user = {
+            'username': 'demo_user'
+        }
+        User.objects.create_user(
```
Please go through this coding style guide https://docs.djangoproject.com/en/2.1/internals/contributing/writing-code/coding-style/ and change this line break according to it.

aditya102 replied:
done

waranlogesh commented on tests/test_user/test_login.py:
```diff
@@ -0,0 +1,57 @@
+            'username': 'demo_user'
+        }
+        User.objects.create_user(
+            username=user['username'], password='demo@123')
```
Give the username directly here. There is no point in creating a separate dictionary just to hold the username which you are gonna use only once.

aditya102 replied:
changed

waranlogesh commented on tests/test_user/test_login.py:
```diff
@@ -0,0 +1,57 @@
+        self.assertEqual(response.status_code, 200)
+        self.assertTemplateUsed(response, 'users/login.html')
+        self.assertContains(response, '<div id="div_id_username" class="form-group"> <label for="id_username" class="col-form-label  requiredField">')
+        self.assertContains(response, '<span class="asteriskField">*</span> </label> <div class=""> <input type="password" name="password" class="textinput textInput form-control" required id="id_password" /> </div> </div')
```
flake8.

waranlogesh commented on tests/test_user/test_login.py:
```diff
@@ -0,0 +1,57 @@
+
+    def test_login_with_invalid_credentials(self):
+        """
+        Test for user is not able to login with invalid credentials
```
It should be `Test for user with invalid credentials.`

waranlogesh commented on tests/test_user/test_login.py:
```diff
@@ -0,0 +1,57 @@
+from django.test import TestCase
```
This file should be `tests/users/test_login_views`

Format is `tests/<app_name>/test_<what_you_are_testing>`

aditya102 replied:
Changes done

waranlogesh commented on tests/test_user/test_login.py:
```diff
@@ -0,0 +1,57 @@
+from django.test import TestCase
+from django.urls import reverse
+
```
This line break should not be there.

aditya102 replied:
changes done

waranlogesh commented on tests/test_user/test_login.py:
```diff
@@ -0,0 +1,57 @@
+        """
+        Test for user is not able to login with invalid credentials
+        """
+        response = self.client.post(reverse('login'), {'username': 'adafafaf', 'password': 'wrong@passowrd'})
```
Testing for one invalid credential is enough.

waranlogesh commented on tests/test_user/test_login.py:
```diff
@@ -0,0 +1,57 @@
+        """
+        response = self.client.post(reverse('login'), {'username': 'adafafaf', 'password': 'wrong@passowrd'})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.context['form'].errors, {'__all__': ['Please enter a correct username and password. Note that both fields may be case-sensitive.']})
```
Use `self.assertFormError` to assert form errors.

aditya102 replied:
done

waranlogesh commented on tests/test_user/test_logout.py:
```diff
@@ -0,0 +1,17 @@
+from django.test import TestCase
```
The file should be under `tests/users/test_logout_views`.

waranlogesh commented on tests/test_user/test_logout.py:
```diff
@@ -0,0 +1,17 @@
+from django.test import TestCase
+from django.shortcuts import reverse
+
```
Remove the empty line here.

waranlogesh commented on tests/test_user/test_logout.py:
```diff
@@ -0,0 +1,17 @@
+
+class AccountsTestCase(TestCase):
+    def setUp(self):
+        user = {
```
Same comment as for the previous test case.

waranlogesh commented on users/views.py:
```diff
@@ -0,0 +1,23 @@
+            return redirect('/users/home')
+    else:
+        form = UserCreationForm()
+    form = UserCreationForm()
```
The above 
```
    else:
        form = UserCreationForm()
```
is not needed.

aditya102 replied:
changed

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,64 @@
+    def test_blankusername_blankpassword(self):
+        response = self.client.post(reverse('login'), {'username': '', 'password': ''})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(
```
Use assertFormError

aditya102 replied:
changed

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,64 @@
+        self.assertEqual(response.status_code, 200)
+        self.assertFormError(response, 'form', None, 'Please enter a correct username and password. Note that both fields may be case-sensitive.')
+
+    def test_validusername_invalid_password(self):
```
test_valid_username_invalid_password

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,64 @@
+        self.assertEqual(response.status_code, 200)
+        self.assertFormError(response, 'form', None, 'Please enter a correct username and password. Note that both fields may be case-sensitive.')
+
+    def test_invalidusername_validpassword(self):
```
test_invalid_username_valid_password

aditya102 replied:
changed

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,64 @@
+        self.assertEqual(response.status_code, 200)
+        self.assertFormError(response, 'form', None, 'Please enter a correct username and password. Note that both fields may be case-sensitive.')
+
+    def test_blankusername_blankpassword(self):
```
test_blank_username_blank_password

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,64 @@
+        self.assertEqual(response.status_code, 200)
+        self.assertFormError(
+            response, 
+            'form', None, 'Please enter a correct username and password. Note that both fields may be case-sensitive.')
```
Have you read [https://docs.djangoproject.com/en/2.1/internals/contributing/writing-code/coding-style/](url) which I posted in my previous PR comments?

aditya102 replied:
done

waranlogesh commented on tests/users/test_logout_views.py:
```diff
@@ -0,0 +1,16 @@
+
+class AccountsTestCase(TestCase):
+    def setUp(self):
+        user = {
```
Remove this

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,64 @@
+        self.assertContains(response, '<input type="password" name="password"')
+        self.assertContains(response, '<button class="btn btn-outline-info" type="submit">')
+
+    def test_post_login_page(self):
```
The name is ambiguous so change it. Also add doc strings to all the test cases.

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,64 @@
+            response.context['form'].errors, {'username': 
+            ['This field is required.'], 'password': ['This field is required.']})
+
+    def test_newpassword_login(self):
```
test_login_after_password_change.

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,64 @@
+
+class AccountsTestCase(TestCase):
+    def setUp(self):
+        User.objects.create_user(username='demo_user', password='demo@123')
```
self.user = User.objects.create_user(username='demo_user', password='demo@123')
and use self.user to access this user throught the test case change it in both test case files.

aditya102 replied:
changes done

waranlogesh commented on tests/users/test_logout_views.py:
```diff
@@ -0,0 +1,16 @@
+        user = {
+            'username': 'demo_user'
+        }
+        User.objects.create_user(username=user['username'], password='demo@123')
```
.

waranlogesh commented on tests/users/test_logout_views.py:
```diff
@@ -0,0 +1,16 @@
+        User.objects.create_user(username=user['username'], password='demo@123')
+
+    def test_logout(self):
+        self.client.login(username='demo_user', password='demo@123')
```
.

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,64 @@
+        self.assertEqual(response.status_code, 200)
+
+    def test_oldpassword_login_check(self):
+        AccountsTestCase.change_password()
```
.

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,64 @@
+            ['This field is required.'], 'password': ['This field is required.']})
+
+    def test_newpassword_login(self):
+        AccountsTestCase.change_password()
```
.

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,64 @@
+        self.assertFormError(response, 'form', None, 'Please enter a correct username and password. Note that both fields may be case-sensitive.')
+
+    def test_validusername_invalid_password(self):
+        response = self.client.post(reverse('login'), {'username': 'demo_user', 'password': '1sfsdf'})
```
.

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,87 @@
+        Helper method for changing password.
+        """
+        new_user = User.objects.get(username='demo_user')
+        new_user.set_password('@new_demo_passowrd')
```
.

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,87 @@
+        """
+        Check login page is redirected to home page or not .
+        """
+        response = self.client.post(reverse('login'), {'username': self.user.username, 'password': 'demo@123'}, follow=True)
```
flake8

waranlogesh commented on tests/users/test_login_views.py:
```diff
@@ -0,0 +1,87 @@
+        """
+        Check for changin password users should be able to login
+        """
+        AccountsTestCase.change_password()
```
.

waranlogesh commented on users/templates/users/home.html:
```diff
@@ -0,0 +1,14 @@
+        <p>You are now log in<p>
+        <p><a href="{% url 'logout' %}">logout</a></p>
+    {% else %}
+        <p>Account Created Succesfully</p>
```
.

waranlogesh commented on tests/users/test_logout_views.py:
```diff
@@ -0,0 +1,13 @@
+    def test_logout(self):
+        self.client.login(username=self.user.username, password='demo@123')
+        response = self.client.get(reverse('logout'), follow=True)
+        self.assertRedirects(response, '/users/login/')
```
Use revers to resolve to this path '/users/login/'

aditya102 commented:
### Add Login and Logout feature 
 - Added Login, Logout and Registration pages.
 - Updated layout with crispy_form.
 - Added test case for Login and Logout.
- Reformated templates pages.
